### PR TITLE
Convert katamari fun/num/obj tests to BDD

### PR DESCRIPTION
**Pull request by techtangents**
Related Ticket: TINY-7469-4

Description of Changes:
* BDD conversions

Pre-checks:
* ~[ ] Changelog entry added~
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/` for new features (if applicable)

Review:
* ~[ ] Milestone set~
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 1]
```diff
@@ -1,141 +1,141 @@
...
+    assert.deepEqual(Fun.curry(c, 'a')(), [ 'a' ]);
+    assert.deepEqual(Fun.curry(c, 'a')('b'), [ 'a', 'b' ]);
+    assert.deepEqual(Fun.curry(c)('a', 'b'), [ 'a', 'b' ]);
+    assert.deepEqual(Fun.curry(c)('a', 'b', 'c'), [ 'a', 'b', 'c' ]);
+    assert.deepEqual(Fun.curry(c, 'a', 'b')('c'), [ 'a', 'b', 'c' ]);
 
-  Assert.eq('eq', false, Fun.not((_x: number) => true)(3));
-  Assert.eq('eq', true, Fun.not((_x: string) => false)('cat'));
+    assert.deepEqual(Fun.not((_x: number) => true)(3), false);
+    assert.deepEqual(Fun.not((_x: string) => false)('cat'), true);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:43]:
```suggestion
    assert.isFalse(Fun.not((_x: number) => true)(3));
    assert.isTrue(Fun.not((_x: string) => false)('cat'));
```

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 2]
```diff
@@ -1,141 +1,141 @@
...
-  called = false;
-  Fun.apply(f);
-  Assert.eq('eq', true, called);
-});
+    let called = false;
+    const f = () => {
+      called = true;
+    };
+    Fun.apply(f);
+    assert.deepEqual(called, true);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:52]:
```suggestion
    assert.isTrue(called);
```

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 3]
```diff
@@ -1,141 +1,141 @@
...
+    called = false;
+    Fun.apply(f);
+    assert.deepEqual(called, true);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:55]:
```suggestion
    assert.isTrue(called);
```

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 4]
```diff
@@ -1,141 +1,141 @@
...
+      let hack: any = null;
+      const output = Fun.call(() => {
+        hack = x;
+      });
 
-    Assert.eq('eq', undefined, output);
-    Assert.eq('eq', x, hack);
-  }));
-});
+      assert.deepEqual(output, undefined);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:134]:
```suggestion
      assert.isUndefined(output);
```

File: modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts [hunk 5]
```diff
@@ -1,68 +1,72 @@
...
+    throttler.throttle('frog');
 
-  setTimeout(() => {
-    Assert.eq('eq', [ 'frog' ], data);
-    throttler.throttle('frog-goose');
-    throttler.throttle('goose');
     setTimeout(() => {
-      Assert.eq('eq', [ 'frog', 'goose' ], data);
-      success();
+      Assert.eq('eq', [ 'frog' ], data);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts:19]: Missed a couple of `Assert` conversions in this file it seems.

File: modules/katamari/src/test/ts/atomic/api/num/NumClampTest.ts [hunk 6]
```diff
@@ -1,19 +1,22 @@
...
+      fc.nat(1000),
+      fc.nat(1000),
+      (a, b, c) => {
+        const low = a;
+        const med = low + b;
+        const high = med + c;
+        // low <= med <= high
+        assert.deepEqual(Num.clamp(med, low, high), med);
+        assert.deepEqual(Num.clamp(med, low, high), med);
+        assert.deepEqual(Num.clamp(high, low, med), med);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumClampTest.ts:19]:
```suggestion
        assert.equal(Num.clamp(med, low, high), med);
        assert.equal(Num.clamp(med, low, high), med);
        assert.equal(Num.clamp(high, low, med), med);
```

File: modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts [hunk 7]
```diff
@@ -1,47 +1,50 @@
...
 
-  Assert.eq('eq', 5, Num.cycleBy(-10, 1, -5, 5));
-  Assert.eq('eq', -5, Num.cycleBy(-10, 50, -5, 5));
+  it(' Unit tests', () => {
+    assert.deepEqual(Num.cycleBy(1, 0, 1, 12), 1);
+    assert.deepEqual(Num.cycleBy(1, 1, 1, 12), 2);
+    assert.deepEqual(Num.cycleBy(1, 2, 1, 12), 3);
+    assert.deepEqual(Num.cycleBy(1, 3, 1, 12), 4);
+    assert.deepEqual(Num.cycleBy(1, 11, 1, 12), 12);
+    assert.deepEqual(Num.cycleBy(1, 12, 1, 12), 1);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts:14]:
```suggestion
    assert.equal(Num.cycleBy(1, 0, 1, 12), 1);
    assert.equal(Num.cycleBy(1, 1, 1, 12), 2);
    assert.equal(Num.cycleBy(1, 2, 1, 12), 3);
    assert.equal(Num.cycleBy(1, 3, 1, 12), 4);
    assert.equal(Num.cycleBy(1, 11, 1, 12), 12);
    assert.equal(Num.cycleBy(1, 12, 1, 12), 1);
```

File: modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts [hunk 8]
```diff
@@ -1,47 +1,50 @@
...
 
-  Assert.eq('eq', 5, Num.cycleBy(50, -150, -5, 5));
-});
+    assert.deepEqual(Num.cycleBy(-10, 1, -5, 5), 5);
+    assert.deepEqual(Num.cycleBy(-10, 50, -5, 5), -5);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts:17]:
```suggestion
    assert.equal(Num.cycleBy(-10, 1, -5, 5), 5);
    assert.equal(Num.cycleBy(-10, 50, -5, 5), -5);
```

File: modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts [hunk 9]
```diff
@@ -1,47 +1,50 @@
...
-    fc.nat(),
-    fc.nat(),
-    (value, delta, min, range) => {
-      const max = min + range;
-      const actual = Num.cycleBy(value, delta, min, max);
-      return (actual - value) === delta || actual === min || actual === max;
-    }
-  ));
-});
+    assert.deepEqual(Num.cycleBy(50, -150, -5, 5), 5);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts:19]:
```suggestion
    assert.equal(Num.cycleBy(50, -150, -5, 5), 5);
```

File: modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts [hunk 10]
```diff
@@ -1,47 +1,50 @@
...
+        const actual = Num.cycleBy(value, delta, min, max);
+        return (actual - value) === delta || actual === min || actual === max;
+      }
+    ));
+  });
+
+  it('0 has no effect', () => {
+    fc.assert(fc.property(fc.nat(), fc.nat(), (value, delta) => {
+      const actual = Num.cycleBy(value, 0, value, value + delta);
+      assert.deepEqual(actual, value);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts:39]:
```suggestion
      assert.equal(actual, value);
```

File: modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts [hunk 11]
```diff
@@ -1,47 +1,50 @@
...
-  fc.assert(fc.property(fc.nat(), fc.nat(), (value, delta) => {
-    const max = value + delta;
-    const actual = Num.cycleBy(value, delta, value, max);
-    Assert.eq('eq', max, actual);
-  }));
+  it('delta is max', () => {
+    fc.assert(fc.property(fc.nat(), fc.nat(), (value, delta) => {
+      const max = value + delta;
+      const actual = Num.cycleBy(value, delta, value, max);
+      assert.deepEqual(actual, max);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/num/NumCycleByTest.ts:47]:
```suggestion
      assert.equal(actual, max);
```

File: modules/katamari/src/test/ts/atomic/api/obj/BiFilterTest.ts [hunk 12]
```diff
@@ -1,62 +1,66 @@
...
-    }
-  ));
-});
+  it('Check that if the filter always returns false, then everything is in "f"', () => {
+    fc.assert(fc.property(
+      fc.dictionary(fc.asciiString(), fc.string(1, 40)),
+      (obj) => {
+        const output = Obj.bifilter(obj, Fun.never);
+        assert.deepEqual(Obj.keys(output.f).length, Obj.keys(obj).length);
+        assert.deepEqual(Obj.keys(output.t).length, 0);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/obj/BiFilterTest.ts:31]:
```suggestion
        assert.lengthOf(Obj.keys(output.f), Obj.keys(obj).length);
        assert.lengthOf(Obj.keys(output.t), 0);
```

File: modules/katamari/src/test/ts/atomic/api/obj/BiFilterTest.ts [hunk 13]
```diff
@@ -1,62 +1,66 @@
...
+    ));
+  });
+
+  it('Check that if the filter always returns true, then everything is in "t"', () => {
+    fc.assert(fc.property(
+      fc.dictionary(fc.asciiString(), fc.string(1, 40)),
+      (obj) => {
+        const output = Obj.bifilter(obj, Fun.always);
+        assert.deepEqual(Obj.keys(output.f).length, 0);
+        assert.deepEqual(Obj.keys(output.t).length, Obj.keys(obj).length);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/obj/BiFilterTest.ts:43]:
```suggestion
        assert.lengthOf(Obj.keys(output.f), 0);
        assert.lengthOf(Obj.keys(output.t), Obj.keys(obj).length);
```

File: modules/katamari/src/test/ts/atomic/api/obj/ObjEachTest.ts [hunk 14]
```diff
@@ -1,31 +1,38 @@
...
+  it('Each + set should equal the same object', () => {
+    fc.assert(fc.property(
+      fc.dictionary(fc.asciiString(), fc.json()),
+      (obj: Record<string, any>) => {
+        const values = {};
+        const output = Obj.each(obj, (x, i) => {
+          values[i] = x;
+        });
+        assert.deepEqual(values, obj);
+        assert.deepEqual(output, undefined);
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/obj/ObjEachTest.ts:34]:
```suggestion
        assert.isUndefined(output);
```

**Review by lnewson — APPROVED** [review]

File: modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts (hunk 5 above)
Comment (techtangents) [reply modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts:19]: Weird.

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 15]
```diff
@@ -1,141 +1,141 @@
-import { Assert, context, describe, it, UnitTest } from '@ephox/bedrock-client';
+import { context, describe, it } from '@ephox/bedrock-client';
 import { assert } from 'chai';
 import fc from 'fast-check';
 import * as Fun from 'ephox/katamari/api/Fun';
 
-UnitTest.test('Fun: unit tests', () => {
-  const add2 = (n) => n + 2;
+describe('atomic.katamari.api.fun.FunTest', () => {
+  it('unit tests', () => {
```
[the remark this answers is not in the document]
Comment (techtangents) [reply modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:7]: Agreed, but I'd prefer that done in a later PR.

File: modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts (hunk 5 above)
Comment (techtangents) [reply modules/katamari/src/test/ts/atomic/api/fun/ThrottlerTest.ts]: Fixed.

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 16]
```diff
@@ -1,141 +1,141 @@
...
+  });
 
-UnitTest.test('Check always :: f(x) === true', () => {
-  fc.assert(fc.property(fc.json(), (json) => {
-    Assert.eq('eq', true, Fun.always(json));
-  }));
-});
+  it('Check always :: f(x) === true', () => {
+    fc.assert(fc.property(fc.json(), (json) => {
+      assert.deepEqual(Fun.always(json), true);
```
Comment (JamesToohey) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:86]:
```suggestion
      assert.isTrue(Fun.always(json));
```

File: modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts [hunk 17]
```diff
@@ -1,141 +1,141 @@
...
+  });
 
-UnitTest.test('Check never :: f(x) === false', () => {
-  fc.assert(fc.property(fc.json(), (json) => {
-    Assert.eq('eq', false, Fun.never(json));
-  }));
-});
+  it('Check never :: f(x) === false', () => {
+    fc.assert(fc.property(fc.json(), (json) => {
+      assert.deepEqual(Fun.never(json), false);
```
Comment (JamesToohey) [inline modules/katamari/src/test/ts/atomic/api/fun/FunTest.ts:92]:
```suggestion
      assert.isFalse(Fun.never(json));
```

**Review by JamesToohey — APPROVED** [review]

**Pull request merged**